kvserver/rangefeed: add basic unbuf reg metrics

wenyihu6 opened the pull request:
**kvserver/rangefeed: add kv.rangefeed.{buffered,unbuffered}_registrations**

This patch introduces two new metrics -
kv.rangefeed.{buffered,unbuffered}_registrations. They track the number of
corresponding active buffered and unbuffered rangefeeds subscribed on the
replica.

Part of: https://github.com/cockroachdb/cockroach/issues/135332
Release note: none

---

**kvserver/rangefeed: add kv.rangefeed.output_loop_nanos**

This patch introduces a new metric, kv.rangefeed.output_loop_nanos, which tracks
the duration of the per-range, per-registration runOutputLoop goroutine. This
goroutine is expected to be short-lived for unbuffered registrations, while it
remains long-lived for buffered registrations. Note that this metric is only
non-zero for unbuffered registrations.

Part of: https://github.com/cockroachdb/cockroach/issues/135332
Release note: none

---

**kvserver/rangefeed: add StreamManagerMetrics**

This patch refactors rangefeed metrics in node.Metrics into a dedicated
StreamManagerMetrics struct. This improves modularity and aligns with future
additions like BufferedSenderMetrics.

Epic: none
Release note: none

---

**kvserver/rangefeed: add rangefeed.buffered_sender.queue_size**

This patch adds a new metric, rangefeed.buffered_sender.queue_size, to track the
number of entries pending in the buffered sender queue waiting to be sent to rpc
stream.

Part of: https://github.com/cockroachdb/cockroach/issues/129816
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add kv.rangefeed.{buffered,unbuffered}_registrations

This patch introduces two new metrics -
kv.rangefeed.{buffered,unbuffered}_registrations. They track the number of
corresponding active buffered and unbuffered rangefeeds subscribed on
replicas.

Part of: #129816
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add kv.rangefeed.output_loop_unbuffered_registration_nanos

This patch introduces a new metric,
kv.rangefeed.output_loop_unbuffered_registration_nanos, which tracks the duration of the
per-range, per-registration runOutputLoop goroutine. This goroutine is expected
to be short-lived for unbuffered registrations, while it remains long-lived for
buffered registrations. Note that this metric is only non-zero for unbuffered
registrations.

Part of: #129816
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137015)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/rangefeed: add StreamManagerMetrics

This patch refactors rangefeed metrics in node.Metrics into a dedicated
StreamManagerMetrics struct. This improves modularity and aligns with future
additions like BufferedSenderMetrics.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add rangefeed.buffered_sender.queue_size

This patch adds a new metric, rangefeed.buffered_sender.queue_size, to track the
number of entries pending in the buffered sender queue waiting to be sent to rpc
stream.

Part of: #129816
Release note: none

wenyihu6 commented:
@stevendanna I remember you mentioned having a few specific metrics in mind that you'd like to include. These are just four basic ones that could provide some initial observability if people want to test this feature.

wenyihu6 commented:
@stevendanna - wanted to revive this PR since it has some metrics we were interested yesterday

stevendanna commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -121,7 +126,8 @@ func (bs *BufferedSender) run(
 			for {
-				e, success := bs.popFront()
+				e, success, remaining := bs.popFront()
+				bs.metrics.BufferedSenderQueueSize.Update(remaining)
```
I don't think you want `Update` here because we might have multiple buffered senders.

wenyihu6 replied:
Ahh, good catch. That explains the unexpected numbers in some of my early runs.

stevendanna commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -223,9 +223,13 @@ func (ubr *unbufferedRegistration) IsDisconnected() bool {
+	defer func() {
+		// Noop if publishCatchUpBuffer below returns no error.
+		ubr.drainAllocations(ctx)
+		ubr.metrics.RangefeedOutputLoopNanos.Inc(timeutil.Since(start).Nanoseconds())
```
If we are only incrementing this for unbuffered registrations. Let's make sure we make that clear in the name and description. 

Also, lets use crtime here just for consistency.

wenyihu6 replied:
Done. While making the changes, I thought it would also be useful to track how long each rangefeed lasts. I opened a separate PR for that since this patch is pretty large - https://github.com/cockroachdb/cockroach/pull/142134.

stevendanna commented on pkg/kv/kvserver/rangefeed/metrics.go:
```diff
@@ -89,6 +101,13 @@ var (
 		Unit:        metric.Unit_COUNT,
 	}
+	metaRangeFeedOutputLoopNanosUnbufferedReg = metric.Metadata{
+		Name: "kv.rangefeed.output_loop_nanos_unbuffered_reg",
```
We usually put the _nanos at the end.

wenyihu6 replied:
Done.

stevendanna reviewed (approved):
Thanks for doing this!

wenyihu6 commented:
TFTR!

bors r=stevendanna

wenyihu6 commented:
blathers backport 25.1
